File: src/agentsareborn/stable.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterable


def utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


class StableStore:
    """Local stable registry for born agents.

    The stable intentionally stores credential references, never raw credentials.
    """

    def __init__(self, root: Path):
        self.root = Path(root)
        self.path = self.root / "agents.json"

    def _empty(self) -> dict:
        return {
            "stableId": "platform-builders",
            "description": "Local stable of platform-builder agents born by AgentsAreBorn.",
            "agents": [],
            "updatedAt": utc_now(),
        }

    def load(self) -> dict:
        if not self.path.exists():
            return self._empty()
        return json.loads(self.path.read_text())

    def save(self, stable: dict) -> None:
        stable["updatedAt"] = utc_now()
        self.root.mkdir(parents=True, exist_ok=True)
        self.path.write_text(json.dumps(stable, indent=2, sort_keys=True) + "\n")

    def add_agent(
        self,
        *,
        agent_id: str,
        name: str,
        role: str,
        manifest_path: str,
        credential_ref: str,
        capabilities: Iterable[str],
        adt_apps: Iterable[str],
        status: str = "embryo",
    ) -> dict:
        stable = self.load()
        agents = [a for a in stable["agents"] if a["agentId"] != agent_id]
        entry = {
            "agentId": agent_id,
            "name": name,
            "role": role,
            "status": status,
            "origin": "agentsareborn",
            "manifestPath": manifest_path,
            "credentialRef": credential_ref,
            "capabilities": list(capabilities),
            "adtApps": list(adt_apps),
            "allowedRunners": ["hermes", "cron", "webhook"],
            "createdAt": utc_now(),
            "lastRunAt": None,
        }
        agents.append(entry)
        stable["agents"] = sorted(agents, key=lambda a: a["name"])
        self.save(stable)
        return entry

    def list_agents(self) -> list[dict]:
        return self.load()["agents"]

File: src/stable.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";

export type PlatformBuilderRole = "propose" | "vote" | "integrate" | "build" | "review" | "release";

export interface StableAgent {
  agentId: string;
  name: string;
  role: PlatformBuilderRole | string;
  status: "embryo" | "active" | "paused" | "retired";
  origin: "agentsareborn";
  manifestPath: string;
  credentialRef: string;
  capabilities: string[];
  adtApps: string[];
  allowedRunners: string[];
  createdAt: string;
  lastRunAt: string | null;
}

export interface StableRegistry {
  stableId: string;
  description: string;
  agents: StableAgent[];
  updatedAt: string;
}

export interface AddAgentInput {
  agentId: string;
  name: string;
  role: PlatformBuilderRole | string;
  manifestPath: string;
  credentialRef: string;
  capabilities: string[];
  adtApps: string[];
  status?: StableAgent["status"];
}

export function utcNow(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

export class StableStore {
  readonly root: string;
  readonly filePath: string;

  constructor(root: string) {
    this.root = root;
    this.filePath = path.join(root, "agents.json");
  }

  private empty(): StableRegistry {
    return {
      stableId: "platform-builders",
      description: "Local stable of platform-builder agents born by AgentsAreBorn.",
      agents: [],
      updatedAt: utcNow(),
    };
  }

  async load(): Promise<StableRegistry> {
    try {
      return JSON.parse(await readFile(this.filePath, "utf8")) as StableRegistry;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        return this.empty();
      }
      throw error;
    }
  }

  async save(stable: StableRegistry): Promise<void> {
    stable.updatedAt = utcNow();
    await mkdir(this.root, { recursive: true });
    await writeFile(this.filePath, `${JSON.stringify(stable, null, 2)}\n`);
  }

  async addAgent(input: AddAgentInput): Promise<StableAgent> {
    const stable = await this.load();
    const agents = stable.agents.filter((agent) => agent.agentId !== input.agentId);
    const entry: StableAgent = {
      agentId: input.agentId,
      name: input.name,
      role: input.role,
      status: input.status ?? "embryo",
      origin: "agentsareborn",
      manifestPath: input.manifestPath,
      credentialRef: input.credentialRef,
      capabilities: [...input.capabilities],
      adtApps: [...input.adtApps],
      allowedRunners: ["hermes", "cron", "webhook"],
      createdAt: utcNow(),
      lastRunAt: null,
    };
    agents.push(entry);
    stable.agents = agents.sort((a, b) => a.name.localeCompare(b.name));
    await this.save(stable);
    return entry;
  }

  async listAgents(): Promise<StableAgent[]> {
    return (await this.load()).agents;
  }

  async recordRun(agentId: string, completedAt = utcNow()): Promise<StableAgent> {
    const stable = await this.load();
    const index = stable.agents.findIndex((agent) => agent.agentId === agentId);
    if (index === -1) throw new Error(`agent not found in stable: ${agentId}`);
    const updated: StableAgent = { ...stable.agents[index], lastRunAt: completedAt };
    stable.agents[index] = updated;
    await this.save(stable);
    return updated;
  }
}
